### PR TITLE
Create LocationValidator for checking new location updates

### LINKED ISSUE #379: Location updates where accuracy is > 100 are ignored
This will be an issue in downtown Chicago, NY etc, where GPS error will be high because of tall buildings. You have to use more intelligent map matching to deal with incoming location updates.

### PR DESCRIPTION
**Pull request by danesfeder**
Closes #379 

Previously, we were checking for horizontal accuracy of `< 100m` and `Location#hasSpeed`.  If the `Location` did not have these requirements, it would be ignored.  

In urban canyons or areas where GPS accuracy really suffers, this could mean we ignore multiple updates at a time (effectively bringing navigation to a halt).

### CODE REVIEW

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LocationValidator.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    log("Accuracy - Current: %s, Previous: %s, Difference: %s",
+      currentAccuracy, previousAccuracy, accuracyDifference);
+
+    boolean currentAccuracyWorse = currentAccuracy > previousAccuracy;
+    boolean hasSameProvider = previousValidLocation.getProvider().equals(location.getProvider());
+    boolean lessThanPercentThreshold = (accuracyDifference <= (previousAccuracy / ACCURACY_THRESHOLD));
+
+    // New location update is acceptable, even with worse accuracy, if it is from
+    // the same provider and is no more than 10% worse
+    boolean worseAccuracyAcceptable = currentAccuracyWorse && hasSameProvider && lessThanPercentThreshold;
```
Comment (devotaaabel) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LocationValidator.java]: maybe reword this to `accuracyAcceptable`? because it doesn't really make sense if `currentAccuracyWorse` is false.

Comment (danesfeder) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LocationValidator.java]: Yeah good point, naming was bad here

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LocationValidator.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      return true;
+    }
+
+    LoggingLocation loggingLocation = new LoggingLocation();
+
+    float currentAccuracy = location.getAccuracy();
+    float previousAccuracy = lastValidLocation.getAccuracy();
+    float accuracyDifference = Math.abs(previousAccuracy - currentAccuracy);
+
+    loggingLocation.setAccuracy(String.format("Accuracy - Current: %s, Previous: %s, Difference: %s",
```
[the remark this answers is not in the document]
Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LocationValidator.java]: BTW, why setting a formatted `String` instead of setting the primitive values into `LoggingLocation`? We can always convert it to `String` afterwards. The same comment applies to `worseAccuracyAcceptable`, `distanceBetweenUpdates`, `timeSinceLastUpdate`, `velocity`, `validVelocity` and `validLocation`.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LocationValidator.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      || timeSinceLastValidUpdate > TIME_THRESHOLD
+      || worseAccuracyAcceptable;
+
+    loggingLocation.setValidLocation(String.format("Valid location: %s", validLocation));
+    sendLoggedLocationUpdate(loggingLocation);
+
+    // Valid velocity and location, set last valid location
+    if (validVelocity && validLocation) {
+      lastValidLocation = location;
+      return true;
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LocationValidator.java]: We could add a variable to `LoggingLocation` to track the final result of the validation. If so, we should call `sendLoggedLocationUpdate` right before exiting.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LoggingLocation.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
+package com.mapbox.services.android.navigation.v5.location;
+
+public class LoggingLocation {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LoggingLocation.java]: Could we rename this DTO? IMO `LoggingLocation` is not reflecting well its purpose.

Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LoggingLocation.java]: BTW, do we want to expose this? I don't 👀 any use cases other than logging.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LocationValidator.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+import android.location.Location;
+import android.support.annotation.NonNull;
+
+import com.mapbox.geojson.Point;
+import com.mapbox.turf.TurfConstants;
+import com.mapbox.turf.TurfMeasurement;
+
+public class LocationValidator {
+
+  private static final int FIVE_SECOND_UPDATE_THRESHOLD = 5000;
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LocationValidator.java]: Minor detail `FIVE_SECONDS_IN_MILLIS_UPDATE_THRESHOLD`

**Comment by kozyr** [comment]
@danesfeder to me it seems this might cause some issues - if your current accuracy is 5, and new accuracy is 10 (still great), wouldn't the code ignore this update as it's not within 10 percent of 5?

**Comment by kozyr** [comment]
@danesfeder and another thing that might happen is something like this (numbers are location accuracies):
5, 10, 10, 10, 10, 10, 500, 500, 10, 10, 10
If you get a location every second, you'd skip all of the 10s, but then when 5 second timer expires for isValidLocation(), it will take 500 without regard to its accuracy, and then it will revert back to 10, which from the user point of view will look like two sudden location jumps, first from one with 5 to 500 (probably inaccurate) and then again back to 10. 

But the proper thing with this sequence would be to ignore both 500 updates. 

Maybe it makes sense to still keep some sorta "it's good enough" location filter that does take all locations under certain threshold (like 50), or look at the last X locations, rather than just the last one?

**Comment by danesfeder** [comment]
@kozyr Yeah that's correct and a good point.  https://github.com/mapbox/mapbox-navigation-android/pull/818 just landed to allow adjustment of these thresholds.
